spacy/tokens/doc.pyx

```diff
@@ -1462,6 +1462,162 @@ cdef class Doc:
+        if token_annotations[HEAD] is not None:
+            token_annotations[HEAD] = [
+                head - i if head is not None else 0 for i, head in enumerate(token_annotations[HEAD])
+            ]
+        if token_annotations[DEP] is not None:
+            deps = []
+            for dep in token_annotations[DEP]:
+                if dep is None:
+                    raise ValueError(Errors)
+                deps.append(dep)
+            token_annotations[DEP] = deps if len(deps) else None
+
+        if token_annotations[DEP] and not token_annotations[HEAD]:
+            token_annotations[HEAD] = [0] * len(token_annotations[DEP])
+        if token_annotations[HEAD] and not token_annotations[HEAD]:
+            raise ValueError(Errors.E1017)
+        if token_annotations[POS] is not None:
+            for pp in set(token_annotations[POS]):
+                if pp not in parts_of_speech.IDS:
+                    raise ValueError(Errors.E1021.format(pp=pp))
+
+        attrs = []
+        annotations = [token_annotations[key] for key in annot_types]
+        possible_headings = [POS, HEAD, DEP, LEMMA, TAG, MORPH]
+        for a, annot in enumerate(annotations):
+            if annot is not None:
+                if len(annot) != len(words):
+                    raise ValueError(Errors.E189)
+                attrs.append(possible_headings[a])
+
+        # If there are any other annotations, set them.
+        if attrs:
+            array = self.to_array(attrs)
+            if array.ndim == 1:
+                array = numpy.reshape(array, (array.size, 1))
+            j = 0
+
+            for annot in annotations:
+                if annot:
+                    if annot is not token_annotations[HEAD]:
+                        for i in range(len(annot)):
+                            if annot[i] is not None:
+                                self.vocab.strings.add(annot[i])
+
+                    if annot is token_annotations[HEAD]:
+                        for i in range(len(words)):
+                            if array.ndim == 1:
+                                array[i] = annot[i]
+                            else:
+                                array[i, j] = annot[i]
+                    elif annot is token_annotations[MORPH]:
+                        for i in range(len(words)):
+                            morph_key = self.vocab.morphology.add(token_annotations[MORPH][i])
+                            if array.ndim == 1:
+                                array[i] = morph_key
+                            else:
+                                array[i, j] = morph_key
+                    else:
+                        for i in range(len(words)):
+                            vocab_key = self.vocab.strings.add(annot[i])
+                            if array.ndim == 1:
+                                array[i] = vocab_key
+                            else:
+                                array[i, j] = vocab_key
```

Comment:
I think this can be simplified to:

```suggestion
                    if annot is token_annotations[HEAD]:
                        for i in range(len(words)):
                            array[i, j] = annot[i]
                    elif annot is token_annotations[MORPH]:
                        for i in range(len(words)):
                            array[i, j] = self.vocab.morphology.add(token_annotations[MORPH][i])
                    else:
                        for i in range(len(words)):
                            array[i, j] = self.vocab.strings.add(annot[i])
```